File: android-app/app/src/main/java/lvt/crm/data/auth/AvatarCrop.kt
package lvt.crm.data.auth

import kotlin.math.max
import kotlin.math.min

object AvatarCrop {
    const val MIN_ZOOM = 1f
    const val MAX_ZOOM = 4f

    fun coverScale(imageWidth: Float, imageHeight: Float, cropSize: Float): Float {
        val smallest = min(imageWidth, imageHeight)
        if (smallest <= 0f || cropSize <= 0f) return 1f
        return cropSize / smallest
    }

    fun clampZoom(zoom: Float): Float = zoom.coerceIn(MIN_ZOOM, MAX_ZOOM)

    fun displayWidth(imageWidth: Float, imageHeight: Float, cropSize: Float, zoom: Float): Float =
        imageWidth * coverScale(imageWidth, imageHeight, cropSize) * clampZoom(zoom)

    fun displayHeight(imageWidth: Float, imageHeight: Float, cropSize: Float, zoom: Float): Float =
        imageHeight * coverScale(imageWidth, imageHeight, cropSize) * clampZoom(zoom)

    fun scale(imageWidth: Float, imageHeight: Float, cropSize: Float, zoom: Float): Float =
        coverScale(imageWidth, imageHeight, cropSize) * clampZoom(zoom)

    fun clampOffset(
        offsetX: Float,
        offsetY: Float,
        displayWidth: Float,
        displayHeight: Float,
        cropSize: Float,
    ): Pair<Float, Float> {
        val minX = min(0f, cropSize - displayWidth)
        val minY = min(0f, cropSize - displayHeight)
        return offsetX.coerceIn(minX, 0f) to offsetY.coerceIn(minY, 0f)
    }

    fun centeredOffset(displayWidth: Float, displayHeight: Float, cropSize: Float): Pair<Float, Float> =
        clampOffset(
            (cropSize - displayWidth) / 2f,
            (cropSize - displayHeight) / 2f,
            displayWidth,
            displayHeight,
            cropSize,
        )

    fun offsetAfterZoom(
        imageWidth: Float,
        imageHeight: Float,
        cropSize: Float,
        oldZoom: Float,
        newZoom: Float,
        offsetX: Float,
        offsetY: Float,
        focusX: Float = cropSize / 2f,
        focusY: Float = cropSize / 2f,
    ): Pair<Float, Float> {
        val oldScale = scale(imageWidth, imageHeight, cropSize, oldZoom)
        val nextZoom = clampZoom(newZoom)
        val nextScale = scale(imageWidth, imageHeight, cropSize, nextZoom)
        if (oldScale <= 0f) {
            return centeredOffset(
                displayWidth(imageWidth, imageHeight, cropSize, nextZoom),
                displayHeight(imageWidth, imageHeight, cropSize, nextZoom),
                cropSize,
            )
        }
        return clampOffset(
            focusX - ((focusX - offsetX) / oldScale) * nextScale,
            focusY - ((focusY - offsetY) / oldScale) * nextScale,
            displayWidth(imageWidth, imageHeight, cropSize, nextZoom),
            displayHeight(imageWidth, imageHeight, cropSize, nextZoom),
            cropSize,
        )
    }

    data class SourceRect(val x: Float, val y: Float, val size: Float)

    fun sourceCropRect(
        imageWidth: Float,
        imageHeight: Float,
        cropSize: Float,
        zoom: Float,
        offsetX: Float,
        offsetY: Float,
    ): SourceRect {
        val drawnScale = scale(imageWidth, imageHeight, cropSize, zoom)
        val size = cropSize / drawnScale
        return clampSourceRect(imageWidth, imageHeight, -offsetX / drawnScale, -offsetY / drawnScale, size)
    }

    fun clampSourceRect(
        imageWidth: Float,
        imageHeight: Float,
        x: Float,
        y: Float,
        size: Float,
    ): SourceRect {
        val maxSquare = min(imageWidth, imageHeight)
        val square = size.coerceIn(1f, maxSquare)
        return SourceRect(
            x.coerceIn(0f, max(0f, imageWidth - square)),
            y.coerceIn(0f, max(0f, imageHeight - square)),
            square,
        )
    }
}

File: src/lib/avatarCrop.js
export const AVATAR_MIN_ZOOM = 1;
export const AVATAR_MAX_ZOOM = 4;
export const AVATAR_OUTPUT_EDGE = 512;

export function coverScale(imageWidth, imageHeight, cropSize) {
  const smallest = Math.min(imageWidth, imageHeight);
  if (!(smallest > 0) || !(cropSize > 0)) return 1;
  return cropSize / smallest;
}

export function clampZoom(zoom) {
  const value = Number(zoom);
  if (!Number.isFinite(value)) return AVATAR_MIN_ZOOM;
  return Math.min(AVATAR_MAX_ZOOM, Math.max(AVATAR_MIN_ZOOM, value));
}

export function displaySize(imageWidth, imageHeight, cropSize, zoom) {
  const scale = coverScale(imageWidth, imageHeight, cropSize) * clampZoom(zoom);
  return {
    width: imageWidth * scale,
    height: imageHeight * scale,
    scale,
  };
}

export function clampOffset(offsetX, offsetY, displayWidth, displayHeight, cropSize) {
  const minX = Math.min(0, cropSize - displayWidth);
  const minY = Math.min(0, cropSize - displayHeight);
  return {
    x: Math.min(0, Math.max(minX, offsetX)),
    y: Math.min(0, Math.max(minY, offsetY)),
  };
}

export function centeredOffset(displayWidth, displayHeight, cropSize) {
  return clampOffset(
    (cropSize - displayWidth) / 2,
    (cropSize - displayHeight) / 2,
    displayWidth,
    displayHeight,
    cropSize,
  );
}

export function offsetAfterZoom({
  imageWidth,
  imageHeight,
  cropSize,
  oldZoom,
  newZoom,
  offsetX,
  offsetY,
  focusX,
  focusY,
}) {
  const nextZoom = clampZoom(newZoom);
  const oldScale = coverScale(imageWidth, imageHeight, cropSize) * clampZoom(oldZoom);
  const next = displaySize(imageWidth, imageHeight, cropSize, nextZoom);
  if (!(oldScale > 0)) {
    return centeredOffset(next.width, next.height, cropSize);
  }
  const anchorX = focusX == null ? cropSize / 2 : focusX;
  const anchorY = focusY == null ? cropSize / 2 : focusY;
  return clampOffset(
    anchorX - ((anchorX - offsetX) / oldScale) * next.scale,
    anchorY - ((anchorY - offsetY) / oldScale) * next.scale,
    next.width,
    next.height,
    cropSize,
  );
}

export function sourceCropRect(imageWidth, imageHeight, cropSize, zoom, offsetX, offsetY) {
  const { scale } = displaySize(imageWidth, imageHeight, cropSize, zoom);
  const size = cropSize / scale;
  return clampSourceRect(imageWidth, imageHeight, {
    x: -offsetX / scale,
    y: -offsetY / scale,
    size,
  });
}

export function clampSourceRect(imageWidth, imageHeight, rect) {
  const maxSquare = Math.min(imageWidth, imageHeight);
  const size = Math.min(Math.max(1, rect.size), maxSquare);
  return {
    x: Math.min(Math.max(0, rect.x), Math.max(0, imageWidth - size)),
    y: Math.min(Math.max(0, rect.y), Math.max(0, imageHeight - size)),
    size,
  };
}
